### PR TITLE
[cxx-interop] Bail on redefined macros.

### PR DESCRIPTION
**Pull request by zoecarver**
This is a stop-gap fix to prevent an infinite loop. Refs SR-13828.

### CODE REVIEW

**Comment by hlopko** [comment]
So, the example test case that you add here is actually an infinite loop, so not importing (and maybe showing a warning) is the right thing to do, imho.

The problem is actually a cycle detection, but your fix only works when the cycle is 1 edge wide. Wdyt about implementing a general solution?

**Comment by zoecarver** [comment]
To prevent this type of recursion, the preprocessor is not supposed to replace a macro with itself. So, the following program is valid:
```
#define TMP_FOO FOO
#undef FOO
#define FOO TMP_FOO
#define OTHER FOO

enum E { OTHER };

void test(E = E::FOO) { }
```

That being said, now that I think about it, there's no case where swift would actually be able to import a plain `FOO` macro, so you're right, we should always bail on this case. I'll update the fix to be more robust (as you suggested).

File: lib/ClangImporter/ImportMacro.cpp [hunk 1]
```diff
@@ -322,15 +322,17 @@ static Optional<std::pair<llvm::APSInt, Type>>
   return None;
 }
 
-
 static ValueDecl *importMacro(ClangImporter::Implementation &impl,
-                              DeclContext *DC,
-                              Identifier name,
-                              const clang::MacroInfo *macro,
-                              ClangNode ClangN,
+                              llvm::SmallSet<StringRef, 4> &visitedMacros,
```
Comment (zoecarver) [inline lib/ClangImporter/ImportMacro.cpp:326]: I couldn't find a `SmallSetImpl` or equivalent so I think we'll just have to use the full `SmallSet<StringRef, 4>`. If you'd prefer I'm happy to add a template parameter to infer the size.

File: test/Interop/C/macros/redefined-macro.swift [hunk 2]
```diff
@@ -0,0 +1,7 @@
+// RUN: %target-swift-ide-test -print-module -module-to-print=RedefinedMacro -I %S/Inputs -source-filename=x | %FileCheck %s
+
+// We cannot import redefined macros.
```
Comment (hlopko) [inline test/Interop/C/macros/redefined-macro.swift:3]: Consider adding some positive tests, testing that we import valid macros.

**Review by hlopko — APPROVED** [review]
LGTM, thank you!

File: test/Interop/C/macros/redefined-macro.swift (hunk 2 above)
Comment (zoecarver) [reply test/Interop/C/macros/redefined-macro.swift:3]: Will do.

**Comment by zoecarver** [comment]
@hlopko I had to change the logic a bit. `name` needs to be the first macro's name because otherwise, we can't create the `ValueDecl` properly. So, I added the `insert` and check right before `importMacro` is called instead of at the top of `importMacro`. An alternative would be to pass the "previous" name as an argument to `importMacro` but that seems more complicated. What do you think?

**Comment by zoecarver** [comment]
@swift-ci please smoke test.

File: lib/ClangImporter/ImportMacro.cpp (hunk 1 above)
Comment (hlopko) [reply lib/ClangImporter/ImportMacro.cpp:326]: I see other instances of such usage in llvm, I think it's fine.

**Review by hlopko — APPROVED** [review]
I'm fine with this, just add an assert assuming the current macro name is in visitedMacros. It's quite possible that somebody will break it in the future (likely me :)

**Comment by zoecarver** [comment]
> I'm fine with this, just add an assert assuming the current macro name is in visitedMacros. It's quite possible that somebody will break it in the future (likely me :)

Do you mean `name`? AFAIK there's nothing equivalent to `macro->getName()` (that's why I have to do the dance of inserting the name before calling `importMacro`).

**Comment by zoecarver** [comment]
@swift-ci please smoke test.

**Comment by zoecarver** [comment]
@swift-ci please smoke test OS X platform.
@swift-ci please smoke test Windows platform.

**Comment by zoecarver** [comment]
@swift-ci please smoke test OS X platform.

**Pull request merged**